Simplenote/src/main/java/com/automattic/simplenote/utils/ReviewAccountVerifier.java

```diff
@@ -0,0 +1,74 @@
+package com.automattic.simplenote.utils;
+
+import android.os.Handler;
+import android.os.Looper;
+
+import com.automattic.simplenote.Simplenote;
+import com.automattic.simplenote.models.Account;
+import com.simperium.client.Bucket;
+import com.simperium.client.BucketObjectMissingException;
+
+import static com.automattic.simplenote.models.Account.KEY_EMAIL_VERIFICATION;
+
+/**
+ * When a user login into their account, Simplenote starts indexing the different buckets including Bucket<Account>.
+ * We setup a OnNetworkChangeListener on the account bucket to listen when the bucket is indexed. One of the parameters
+ * of the callback onNetworkChange called is Bucket.ChangeType. If this parameter is INDEX, then we know that the
+ * account bucket was indexed (synchronized). When the account bucket is synchronized, we can check whether the account
+ * was verified or not
+ */
+public class ReviewAccountVerifier implements Bucket.OnNetworkChangeListener<Account> {
+    private enum Status {
+        UNKNOWN,
+        UNVERIFIED,
+        VERIFIED,
+    }
+
+    private final Simplenote simplenote;
+    private Status currentStatus = Status.UNKNOWN;
+
+    public ReviewAccountVerifier(Simplenote simplenote) {
+        this.simplenote = simplenote;
+    }
+
+    public void clearStatus() {
+        currentStatus = Status.UNKNOWN;
+    }
+
+    @Override
+    public void onNetworkChange(Bucket<Account> bucket, Bucket.ChangeType type, String key) {
+        // If we already know the answer, we avoid computing it again
+        if (currentStatus != Status.UNKNOWN) {
+            return;
+        }
+
+        String email = simplenote.getUserEmail();
+        if (type != Bucket.ChangeType.INDEX || email == null) {
+            return;
+        }
+
+        // When a network change of type INDEX is received, it means that the channel finished indexing the bucket, thus
+        // we can check whether the account is verified or not
+        try {
+            Account account = bucket.get(KEY_EMAIL_VERIFICATION);
+            boolean hasVerifiedEmail = account.hasVerifiedEmail(email);
+            if (!hasVerifiedEmail) {
+                currentStatus = Status.UNVERIFIED;
+
+                final boolean hasSentEmail = account.hasSentEmail(email);
+                Looper looper = Looper.getMainLooper();
+                Handler handler = new Handler(looper);
+                handler.post(new Runnable() {
+                    @Override
+                    public void run() {
+                        simplenote.showReviewVerifyAccount(hasSentEmail);
+                    }
+                });
+            } else {
+                currentStatus = Status.VERIFIED;
+            }
+        } catch (BucketObjectMissingException e) {
+            AppLog.add(AppLog.Type.SYNC, "Account for email verification is missing");
+        }
```

Comment:
superficially here it'd be nice if we limited the `try`/`catch` to surrounding the statements that could crash, also of early-aborting where possible to avoid needless nesting. in this case it would be something like the following…

```java
Account account;
try {
    account = bucket.get(KEY_EMAIL_VERIFICATION);
} catch (BucketObjectMissingException e) {
  AppLog.add(…);
  return;
}

if (account.hasVerifiedEmail(email)) {
  currentStatus = Status.VERIFIED;
  return;
}

currentStatus = Status.UNVERIFIED;
…
```

Reply:
on a deeper level I think there are a few challenges with this section of code and the way we're coupling the behavior of the verification monitor to the broader application, especially the `looper`/`handler` lines.

what we've done is push the implementation of the app's logic into this helper class and in doing so overlooked an opportunity to abstract: to define some new names of terms and behaviors. if we go back and ask ourselves what we're want to do, essentially, then I think it might lead us to a few changes that would clear up the code _and_ decouple these bits.

if I can reference `SyncTimes` again we did the same kind of thinking there and [created a new interface](https://github.com/Automattic/simplenote-android/blob/68efd052b29eb88a9e2bbe4476700d91acab5a5e/Simplenote/src/main/java/com/automattic/simplenote/SyncTimes.java#L122-L125) for parent/host/calling classes to implement. by doing this and keeping the internal details of the modules decoupled we retain control in "the caller" (in this case, the `Simperium` class).

```java
interface VerificationStateListener {
    void onUpdate(Status status);
}
```

if we think about the verification status though `onUpdate` we can see that we don't actually need to export boolean states for `hasSentEmail` or probably even track `mHasShownReviewOrVerify`. let's consider a subtle change in our `Status` enum to focus on the verification process more than the data in the `account` bucket.

```java
public enum Status {
    Verified,
    SentEmail,
    Unverified
}
```

as far as we care, we only need to know these states. we can track `Unknown` but it's also a value we don't need to export, so maybe it's possible to use a `null` for the internal status tracker inside our monitor. I don't know, `Unknown` isn't a bad value.

```java
// Simplenote.java

mVerificationWatcher = new AccountVerificationWatcher(this);
mAccountBucket.addOnNetworkChangeListener(mVerificationWatcher.bucketListener);
mVerificationWatcher.addListener(this.verificationListener);

private class VerificationListener implements AccountVerificationWatcher.VerificationStateListener {
    @Override
    public void onUpdate(AccountVerificationWatcher.Status status) {
        switch (status) {
            case VERIFIED:
                closeNag();
                return;

            case SENT_EMAIL:
                showWaitingOnEmailConfirmation();
                return;

            case UNVERIFIED:
                showUnverifiedWarning();
                return;
        }
    }
}
```

it might seem verbose, but it lets us treat each class independently and it removes our need to track state. we know that `onUpdate` won't be called unless the status _changes_ and that also means it won't be called until we get our first update from the server.

with this alternate design I think it will be clearer to watch the flow of data in the bucket and make it easier not to introduce another bug because of updates coming across after (or before) indexing operations complete.

Reply:
> superficially here it'd be nice if we limited the try/catch to surrounding the statements that could crash, also of early-aborting where possible to avoid needless nesting. in this case it would be something like the following…

I personally don't like this style because it becomes too choppy. The first check `if (type != Bucket.ChangeType.INDEX || email == null)` we cut it earlier because it is more a constraint, but if we split the algorithm in parts, IMO It makes the logic behind the method harder to understand. So, I prefer the method to flow as it is since it just has two paths (low cyclomatic complexity)

Reply:
> on a deeper level I think there are a few challenges with this section of code and the way we're coupling the behavior of the verification monitor to the broader application, especially the looper/handler lines.

One way we can solve this in a simple way is moving the handler code inside `Simplenote` which has the knowledge about the current activity and can use the activity to call the method `runOnUIThread` instead. However, I see your previous comments and this one have some concerns about the whole architecture and coupling. Let me address those comments before making this change.

Reply:
> I personally don't like this style because it becomes too choppy.

That's fair. I think our Android code is full of nesting but that's uncommon in our other projects. Most of the time we prefer early-aborting to avoid nesting.

> if we split the algorithm in parts, IMO It makes the logic behind the method harder to understand

just be careful not to make adverse decisions because of some change we might make in the future but have no plans to make. there are ways of handling changes when they come, in a case like this many times we find that instead of adding more nesting levels, what we end up discovering is that we have a dispatcher and multiple methods to handle each dispatch.

> So, I prefer the method to flow as it is right now since it just has two paths (low cyclomatic complexity)

This nested structure has higher cyclomatic complexity three paths. This is why I like avoiding the nesting and early-aborting, because the control flow isn't deceiving as it can be in cases like this 😉. We know that once we hit that `return` statement we're done. Here, we have to remember that the `try`/`catch` interacts with the conditionals.

If we narrow the `try`/`catch` so that it only wraps `bucket.get()` then we avoid increasing the cyclomatic complexity but we do have to remember to `return` from the `catch` clause or else we'll continue to run the verification check even if the entity is missing.

> makes the logic simpler

I'm not sure this true. It is probably more familiar to your style, but in order to simply the logic we have to remove branches or variables. in this case I'm pretty sure it's mostly a style preference apart from the fact that we have to remember to skip to the end of the `try`/`catch` to see if there's any further processing in the method (whereas with a `return` we can know without looking any further, "this is where execution ends.")

> One way we can solve this in a simple way is moving the handler code inside Simplenote which has the knowledge about the current activity and can use the activity to call the method runOnUIThread instead. 

If that means adding the listener code in `Simperium.java` then that's definitely not what I'm recommending. I like that you created a separate class/object to watch the verification status and that is a positive move for the project. We're already way too-coupled in places and our classes typically do way too much. Isolating behaviors into independent and semantically-narrow classes is a good thing here.

> concerns about the whole architecture and coupling. Let me address those comments before making this change.

👍 awesome. I think that with some rethinking of the `AccountVerificationWatcher` interface we will end up in an easier spot; we won't have to answer some of the current questions about architecture because they will be solved by construction for us.

Reply:
> If that means adding the listener code in Simperium.java then that's definitely not what I'm recommending.

No, this is not what I meant. But is futile to discuss it if we are going to change the architecture. 

> We know that once we hit that return statement we're done. Here, we have to remember that the try/catch interacts with the conditionals.

We could talk ages about style, but in the end, it is getting subjective since we do not have a cognitive study about 🤣 . 

> We're already way too-coupled in places and our classes typically do way too much. Isolating behaviors into independent and semantically-narrow classes is a good thing here.

I definitely agree with your sentiment and I think the architecture you presented for account verification is great (let me prepare the changes, I will have them shortly). I just want to be careful of applying half-patches improvements that when if thought better, we could make something really simpler with higher quality. For example, you mentioned after logout we should not make all those calls manually (they should be part of the process itself). In that case, we could go a try to move that code, but if we focus on designing a holistic solution for session management, we might end up with a better solution. There are some spots that need some work and I really appreciate that from this small PR you have identified more than two spots for improvements. But I think we could handle those cases in separate PRs.

Reply:
> We could talk ages about style, but in the end, it is getting subjective since we do not have a cognitive study to support our claims 🤣 .

Note that I'm not saying one is better than the other. It's not a subjective claim I'm making: I'm simply pointing out that with early-abort patterns we are able to safely turn our minds off _at the point where we return_ because it's impossible for there to be continuing behaviors whereas with the nesting we aren't allowed to do that - we have to continue scanning until the end of the block to check if there is more code that might still run.

> I just want to be careful of applying half-patches improvements that when if thought better, we could make something really simpler with higher quality.

I'm not talking about restructuring the app in this PR. I'm saying instead that we should not feel bound to a pattern that's prevalent in the app which has caused headaches, that we are free to make new things with a better design than existing things have.

What are the half-patches you're talking about?

> For example, you mentioned after logout we should not make all those calls manually (they should be part of the process itself).

Specifically in that comment I mentioned that it's not work for this PR since it involves a larger redesign.

> In that case, we could go a try to move that code, but if we focus on designing a holistic solution for session management, we might end up with a better solution. There are some spots that need some work and I really appreciate that from this small PR you have identified more than two spots for improvements. But I think we could handle those cases in separate PRs.

I'm confused on what you're attempting to say here. Can you reword this or maybe be more specific about what you're talking about? In my own suggestions I'm _only_ talking about the design points of the account verification watcher and not of changing anything else in the app. If I've communicated doing broader work as a requirement here that was me miscommunication; I mention parts of the app as _motivation_ for design on this change, not as a dependency before it.

Reply:
> I'm not talking about restructuring the app in this PR. I'm saying instead that we should not feel bound to a pattern that's prevalent in the app which has caused headaches, that we are free to make new things with a better design than existing things have.

Agree. I think the improvements you presented leave the PR in a better position from a design point of view.

> In my own suggestions I'm only talking about the design points of the account verification watcher and not of changing anything else in the app.
> you don't need to feel bound to the style comments. while I'm not going to ignore the observation or discussion (I'm happy to hold that conversation) it's not central to the needs of this PR.

Yes, I think this is the part the confuses me a little bit about your review. I think I made them too much part of the conversation which got mixed with the rest of the comments that define the behavior and better architecture.